### PR TITLE
Reduce InternalComposite in a streaming fashion

### PR DESCRIPTION
**Pull request by iverase**
We are currently using a merge sort algorithm to merge buckets on InternalComposite. The issue with this algorithm is that it requires to have all buckets to be merged on heap which is inefficient. This PR proposes to only keep the competitive aggregations in memory.

In order to do that be use a priority queue with a hashmap to keep traclk of the competitive buckets. We still delayed the merging of the child aggregations by introducing a DelayedMultiBucketAggregatorsReducer.

I did compare the algorithms by merging 500 shards with 65k different values in 2gib heap nodes. The results are encouraging, the current algorithm will go out of memory for composite sizes >= 30k:

```
size 1000 : 61.26956375
size 10000 : 39.356867625
size 20000 : 40.178292459000005
size 30000 : * node crash *
```

With the new algorithm, the process is twice as fast and does not go out of memory:

```
size 1000 : 31.356643000000002
size 10000 : 20.6347565
size 20000 : 17.899176792000002
size 30000 : 18.150707125
size 40000 : 18.304503375
size 50000 : 18.923937417
size 65536 : 20.001164250000002
```

relates https://github.com/elastic/elasticsearch/pull/105207

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/DelayedMultiBucketAggregatorsReducer.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+     */
+    public void reset() {
+        count = 0L;
+        internalAggregations.clear();
+    }
+
+    /**
+     * returns the reduced {@link InternalAggregations}.
+     */
+    public InternalAggregations get() {
```
Comment (nik9000) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/DelayedMultiBucketAggregatorsReducer.java:58]: s/get/reduce/ ?

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/DelayedMultiBucketAggregatorsReducer.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+            }
+            return aggregatorsReducer.get();
+        }
+    }
+
+    /**
+     * returns the number of docs
+     */
+    public long getDocCount() {
+        return count;
```
Comment (nik9000) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/DelayedMultiBucketAggregatorsReducer.java:71]: It has an accessor method so maybe it shouldn't be package private?

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/InternalComposite.java [hunk 3]
```diff
@@ -275,11 +251,82 @@ public InternalAggregation get() {
...
+                    // bucket is added
+                    delayed = new DelayedMultiBucketAggregatorsReducer(reduceContext);
+                } else if (out == bucket) {
+                    // bucket is not competitive
+                    return false;
+                } else {
+                    // bucket replaces existing bucket
+                    delayed = bucketReducers.remove(out.key);
+                    assert delayed != null;
+                    delayed.reset();
```
Comment (nik9000) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/InternalComposite.java:290]: Ah! Now I see why you reset it.

**Review by nik9000 — APPROVED** [review]
Very neat.

**Review by nik9000 — COMMENTED** [review]
When you apply this pattern to `terms` you'll have to merge *some* of the sub-buckets so you can perform sorting on them. That'll be fun.

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/DelayedMultiBucketAggregatorsReducer.java (hunk 2 above)
Comment (iverase) [reply server/src/main/java/org/elasticsearch/search/aggregations/bucket/DelayedMultiBucketAggregatorsReducer.java:71]: yes indeed

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/DelayedMultiBucketAggregatorsReducer.java (hunk 1 above)
Comment (iverase) [reply server/src/main/java/org/elasticsearch/search/aggregations/bucket/DelayedMultiBucketAggregatorsReducer.java:58]: My next PR is going to refactor this classes a bit. My idea is that they can look very similar to a bucket (and therefore we can apply BucketOrder on it).I keep this for that next PR.

**Comment by iverase** [comment]
>When you apply this pattern to terms you'll have to merge some of the sub-buckets so you can perform sorting on them. That'll be fun.

Yes, the idea is that the buckets reducer can implement the buckets interface (so ti would need to be more similar to DelayableBucket) and therefore we can apply BucketOrder on them.

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/DelayedMultiBucketAggregatorsReducer.java (hunk 1 above)
Comment (nik9000) [reply server/src/main/java/org/elasticsearch/search/aggregations/bucket/DelayedMultiBucketAggregatorsReducer.java:58]: Enjoy

**Pull request merged**